entity-service-client/src/main/java/org/hypertrace/entity/data/service/client/EdsCacheClient.java

```diff
@@ -224,4 +226,71 @@ public void upsertEnrichedEntities(String tenantId, EnrichedEntities enrichedEnt
   public void upsertRelationships(String tenantId, EntityRelationships relationships) {
     client.upsertRelationships(tenantId, relationships);
   }
+
+  public void updateBasedOnChangeEvent(
+      EntityChangeEventKey entityChangeEventKey, EntityChangeEventValue entityChangeEventValue) {
+    LOG.debug("Entity change event is {}, {} ", entityChangeEventKey, entityChangeEventValue);
+
+    switch (entityChangeEventValue.getEventCase()) {
+      case CREATE_EVENT:
+        // ignore create events, don't populate caches if not necessary
+        break;
+      case UPDATE_EVENT:
+        updateCacheValues(
+            entityChangeEventKey, entityChangeEventValue.getUpdateEvent().getLatestEntity());
+        break;
+      case DELETE_EVENT:
+        getEntityCacheKeys(entityChangeEventKey)
+            .forEach(cacheKey -> entityCache.invalidate(cacheKey));
+        entityIdsCache.invalidate(
+            getIdsCacheKey(
+                entityChangeEventKey, entityChangeEventValue.getDeleteEvent().getDeletedEntity()));
+        break;
+      default:
+        LOG.warn(
+            "Entity change event value has invalid event type -> {}",
+            entityChangeEventValue.getEventCase());
+    }
+  }
+
+  private void updateCacheValues(EntityChangeEventKey entityChangeEventKey, Entity entity) {
+    getEntityCacheKeys(entityChangeEventKey)
+        .forEach(
+            cacheKey -> entityCache.asMap().computeIfPresent(cacheKey, (key, oldEntity) -> entity));
```

Comment:
sorry I missed before, but rather than converting the cache to a map and directly modifying it (two different steps that rely on understanding documented but unintuitive impl - that the map is live and modifying it modifies the cache, and that computeIfPresent has side effects that update the map) suggest using cache.put to modify the state (after doing a contains check, if needed). Among other things, I'm not sure how modify the map directly impacts things like TTLs and metrics (and whether it handles correctly or not is almost besides the point since it's not obvious one way or the other).